velox/exec/HashBuild.cpp

```diff
@@ -178,12 +277,316 @@ void HashBuild::addInput(RowVectorPtr input) {
+  }
+  const auto& hashers = table_->hashers();
+  for (auto i = 0; i < hashers.size(); ++i) {
+    auto& hasher = hashers[i];
+    if (hasher->channel() != kConstantChannel) {
+      hashers[i]->hash(activeRows_, i > 0, hashes_);
+    } else {
+      hashers[i]->hashPrecomputed(activeRows_, i > 0, hashes_);
+    }
+  }
+
+  spillPartitions_.resize(input->size());
+  for (auto i = 0; i < spillPartitions_.size(); ++i) {
+    spillPartitions_[i] = spiller_->hashBits().partition(hashes_[i]);
+  }
+}
+
+void HashBuild::spillPartition(
+    uint32_t partition,
+    vector_size_t size,
+    const BufferPtr& indices,
+    const RowVectorPtr& input) {
+  VELOX_DCHECK(spillEnabled());
+
+  if (isInputFromSpill()) {
+    spiller_->spill(partition, wrap(size, indices, input));
+  } else {
+    spiller_->spill(
+        partition,
+        wrap(size, indices, tableType_, spillChildVectors_, input->pool()));
+  }
+}
+
+bool HashBuild::requestSpill(RowVectorPtr& input) {
+  VELOX_CHECK_GT(numSpillRows_, 0);
+  VELOX_CHECK_GT(numSpillBytes_, 0);
+
+  input_ = std::move(input);
+  if (spillGroup_->requestSpill(*this, future_)) {
+    VELOX_CHECK(future_.valid());
+    setState(State::kWaitForSpill);
+    return false;
+  }
+  input = std::move(input_);
+  return true;
+}
+
+bool HashBuild::waitSpill(RowVectorPtr& input) {
+  if (!spillGroup_->needSpill()) {
+    return false;
+  }
+
+  if (spillGroup_->waitSpill(*this, future_)) {
+    VELOX_CHECK(future_.valid());
+    input_ = std::move(input);
+    setState(State::kWaitForSpill);
+    return true;
+  }
+  return false;
+}
+
+void HashBuild::runSpill(const std::vector<Operator*>& spillOperators) {
+  VELOX_CHECK(spillEnabled());
+  VELOX_CHECK(!spiller_->state().isAllPartitionSpilled());
+
+  uint64_t targetRows = 0;
+  uint64_t targetBytes = 0;
+  std::vector<Spiller*> spillers;
+  spillers.reserve(spillOperators.size());
+  for (auto& spillOp : spillOperators) {
+    HashBuild* build = dynamic_cast<HashBuild*>(spillOp);
+    VELOX_CHECK_NOT_NULL(build);
+    spillers.push_back(build->spiller_.get());
+    build->addAndClearSpillTarget(targetRows, targetBytes);
+  }
+  VELOX_CHECK_GT(targetRows, 0);
+  VELOX_CHECK_GT(targetBytes, 0);
+
+  std::vector<Spiller::SpillableStats> spillableStats(
+      spiller_->hashBits().numPartitions());
+  for (auto* spiller : spillers) {
+    spiller->fillSpillRuns(spillableStats);
+  }
+
+  // Sort the partitions based on the amount of spillable data.
+  SpillPartitionNumSet partitionsToSpill;
+  std::vector<int32_t> partitionIndices(spillableStats.size());
+  std::iota(partitionIndices.begin(), partitionIndices.end(), 0);
+  std::sort(
+      partitionIndices.begin(),
+      partitionIndices.end(),
+      [&](int32_t lhs, int32_t rhs) {
+        return spillableStats[lhs].numBytes > spillableStats[rhs].numBytes;
+      });
+  int64_t numRows = 0;
+  int64_t numBytes = 0;
+  for (auto partitionNum : partitionIndices) {
+    if (spillableStats[partitionNum].numBytes == 0) {
```

Comment:
Hi @xiaoxmeng , why directly break if one partition's bytes are zero? It's possible that the following partitions aren't empty, no?